### ISSUE TITLE
Error loading from character file: Character version is not compatible

### ISSUE THREAD

**Issue by Knoopington** [Wufflez/Loki#39]

**Comment by gahstanz** [comment]
Few questions.
1. Did you recently update your client?
2. What version is your current client?
3. Can you please provide a sample of the file that is causing the issue?

**Comment by assBuk** [comment]
The version of the client and the game is most likely the latest ¯\_(ツ)_/¯

**Comment by Ecksters** [comment]
@jensbrak To my knowledge yours was the only version of Loki that supported post-Mistlands update. However, it seems like the Hildir's Request update has broken the character editor again, any chance you'd be willing to fix it since Wufflez seems unavailable?

**Comment by jensbrak** [comment]
> However, it seems like the Hildir's Request update has broken the character editor again

I noticed that and started look at the changes. It's more than a few new items, the save format has been quite extensively updated. Nothing advanced but more than the usual tweaks for sure (new stats system and some updates to the skill system too IIRC).

> any chance you'd be willing to fix it since Wufflez seems unavailable?

I'll see what I can do. Nothing particular complicated but a little more to test. Not sure if I will extend my branch of the program or the main one. I'll give a heads up here if I get the time to do an update. If no one beats me to it that is.

Other than that I've played some more with Unity and dived a little deeper in the Valheim code base (already made my first mod. Nothing worthy a release but it's even more fun to fool around with the code at runtime than with the save files.)

**Comment by jensbrak** [comment]
A little update: just as I had made the changes in (my version of) Loki to make it work with the Hildir update, Valheim was updated once again. And of course it included changes affecting how characters are loaded and saved. I have now updated (again my version of) Loki once again. It seems to work but needs a bit consideration and testing I feel. Why?

The main problem is it seems the official Loki version is not actively maintained, or so it seems. My version of Loki might be the latest one updated to be compatible with Valheim as it has changed, but I have no intention of becoming the maintainer of Loki. As Valheim is evolving (with respect to character/player data), it becomes more and more difficult to "just throw in" things into Loki without having to do design choices and more. 

I'll see if I can redo things but base it on the official Loki code instead of my branch with its own life (sigh). If not, I'll just complicate things even more. I made my branch to toy with Loki, not with the intention to change it in the long run.

(Since I've learned loads of stuff about Unity in general and Valheim in particular since I first started to toy with Loki, I now can give pointers for anyone wanting to modify their characters. Each solution with their pros and cons respectively).

**Comment by Ecksters** [comment]
@jensbrak Thanks for your work on this, I totally understand not wanting to take over maintenance of the project. Is there a different tool that you think is preferable for making these modifications, such as an in-game mod? I just tend to like restarting when new content is added, but being able to skip the resource grind a bit makes it more enjoyable.

I just found this: https://www.nexusmods.com/valheim/mods/1536/

And an in-game one: https://www.nexusmods.com/valheim/mods/104

Seems to be being actively updated, maybe that's worth recommending over Loki.

**Comment by jensbrak** [comment]
I will update either my branch or make a new branch more compatible with the original version and put the changes there.  Just because it would be a waste of effort not to since I've done the changes in question. Just have to decide which way to go.  

(If Loki is to be considered not longer being maintained, it really doesn't matter where the changes go. If someone (Loki creator or anyone else) picks up the development, it would be unfortunate to have changes done in my branch of it. Simply because my branch contain both proposed new features (not merged into official version) as well as updates to maintain compability with Valheim. In the end the changes are not that big, really.)

To your question about not having to grind: totally with you there. :) The answer kind of depend on how you play Valheim. For worlds you've created yourself, the easiest and certainly most reliable option is to use in game console commands. They unlock pretty much anything the game itself support. Only caveat is you cannot use these commands if any other player join your world (at least not while they're in it). Look up enable console commands in Valheim and you're soon able to spawn inventory items/stacks and repair them "for free" inside the game. Be warned though, while it unlocks unlimited powers (great for epic builds like those show cases on gaming sites) it also risks "nerfing the game" for you in the end (depending on how what extent you use them). I think it was a wise decision to limit those commands for self hosted single play only.

Mods are another path to go. I've actually never tried any premade mods. Instead, I jumped directly into the rabbit hole of  creating mods myself. Just out of curiosity. But I do know there are sites for installing and updating mods within Valheim (and other games). I'd look for Thunderstone to begin with. Or, as you mentioned, Nexusmods. The benefit of mods are that they enhance the game, in the game itself. The downside is that they need to be maintained by the devs and updated when Valheim is updated. As with Loki and other off game tools. I have no specific recommendation for a mod. Many mods have their source code published on GitHub too, like Loki has. That's how I learned about them.

TL;DR - Single play self hosted worlds: enable and use console commands. Or finds mods on Thunderstone that might do the trick. But for that isolated, pretty safe update of a character save with no other implications than having to backup save files, Loki was the best option I could find.

**Comment by jensbrak** [comment]
@Ecksters (and anyone else missing a working version of Loki): 

I have made a custom release of Loki that works with the latest (at the time of writing) version of Valheim (0.217.24, Hildir update). It's available in my fork of Loki and released as Loki version 0.13-beta (aka Phivoc Hildir Update).

_Please read the release notes of it carefully before downloading and using it._ 
Link: https://github.com/jensbrak/Loki/releases/tag/v0.13-beta

**Comment by jensbrak** [comment]
Just curious. Would it be of any interest with another update? 

I didn't stop after releasing the update. Mostly out of curiosity (of the answer to "how hard could it be?") I added some features and while they do work just fine already, it would require some small tidying up if anyone else would be interested. Right now I have a version with:

- Button to repair ALL worn items in inventory, with one click (rather than doing it manually on all items in need of repair)
- Button to fill ALL stacks in inventory, with one click (applies to all stackable items not full, from foods and materials to trophies  and money)
- Button to modify ALL skills by a selectable percentage (right now limited to be -100 to +100%) where default is increase by +5% (which "happen" to match the amount skills are lowered by death, default setting that is)
- A new Stats tab with ALL the new stats provided with the new Valheim save format. Non interactive to begin with, after all they're just stats, right? :)

There are other things that could be added quite easily (like "the rest" of the new stuff in the new save format) but I stopped with the above. After all, I don't use Loki for playing normally - these were just the things that I felt was missing the most. 

![image](https://github.com/Wufflez/Loki/assets/52660827/ea379bd4-7d6c-466b-89f5-fb6fb52f61e7)
![image](https://github.com/Wufflez/Loki/assets/52660827/b5abfb12-0bf2-4e6f-93b7-b0e2d03e29ce)
![image](https://github.com/Wufflez/Loki/assets/52660827/773b7f1c-34bd-4802-904a-1775770b9ff6)

**Comment by Ecksters** [comment]
Haha, that's awesome, I'm still using it for now, so I certainly wouldn't complain, but you've brought to my attention that there are other tools, so certainly don't feel compelled to keep maintaining it.

**Comment by jensbrak** [comment]
Allrighty then! It's just a matter of cleaning up some test code and make a new release on github with what I already have. I wouldn't have asked if I wasn't ready to do it. After diving in to the rabbit hole of creating mods and reverse engineering unity code and content, I realize how much I appreciate the isolated world of just managing save data.  Both are great fun but have their pros and cons respectively :)

**Comment by jensbrak** [comment]
@Ecksters woop woop!

https://github.com/jensbrak/Loki/releases/tag/v0.14-beta

**Comment by jensbrak** [comment]
If anyone still use this program, I have just updated my fork to work with Ashlands update (that came today).
From the looks of it, only minor changes have been made to character save file format, it seems to work fine.
However, I want to test it some more first before considering a release.

**Comment by jensbrak** [comment]
Updated my fork and made a new release with Ashlands support.
At least somewhat tested but there could be things missing/wrong, let me know.
As usual, read release notes before considering using it.

https://github.com/jensbrak/Loki/releases/tag/v0.15-beta

**Comment by DarkJellyfish** [comment]
@jensbrak Thanks for the update! I couldn't comment on your release itself, but I've played around with it a bit and no major issues so far. No crashes in game, inventory/skills are as expected, files load just fine, backups work.

One small peculiarity - when using the .exe to open, characters load fine into the dropdown, everything works great. When opening the app with the character file (right click the file, open with, choose the .exe) the inventory doesn't load properly: item names are missing and nothing can be changed. Just something odd I stumbled upon. 

Again, this is great, thank you.

**Comment by jensbrak** [comment]
> One small peculiarity - when using the .exe to open, characters load fine into the dropdown, everything works great. When opening the app with the character file (right click the file, open with, choose the .exe) the inventory doesn't load properly: item names are missing and nothing can be changed. Just something odd I stumbled upon.

@DarkJellyfish Interesting! I haven't really tried to open a file that way (open with) but my hunch is that this has not worked even before I started to mess with the code. But I can be wrong. I'll have a look on it, not because I feel obliged, more like intrigued. 

Glad it works, nice to get some feedback too!

**Comment by jensbrak** [comment]
@DarkJellyfish Just had a look. It is actually my code that is to blame. When separating the item database to a separate file, I made an assumption that seems to be false if starting Loki using "Open with..." (a case I didn't consider). The fix is super easy and will make inventory look as expected. I'll publish that fix at some time!

A note though: Loki is NOT built to take "Open with..." into account.  In other word, original author did not consider this case either, as far as I can see: If you launch Loki with a right click on a .fch-file, Loki will discard this information. It will instead load ALL files within a predetermined directory.  Which likely is the same directory where the file you right clicked on is, making it feel like the right click actually loaded that file into Loki as a result. 

You can easily verify this: copy an .fch-file to a completely different location (say Desktop), right click and choose Open with. Loki will open a number of files. But look closely in the paths of each file (within Loki - under the title of the save files). The one you right clicked on (on the Desktop) is not there. But likely the original copy of the one you copied.

Even this can be fixed but is more work to do. Mainly because the intention of Loki is to open all files within the save dir only. Adding the file being right clicked is easy. But then you also might wanna skip the loading of the files in the save dir. Or combine those results (with the risk of user mixing up the version of the file in the save dir with a specific copy of it elsewhere). Bottom line: design choices affecting usability will have to be made. Implementation is rather easy though.

**Comment by jensbrak** [comment]
In case anyone wonder, yes, I have adjusted Loki to work with the 0.219.13 "The Bog Witch" version (which was released a couple of days ago).
Also added possibility to explicitly open a .fch file anywhere  with Loki using "open with..." (as mentioned in previous answers here). This wasn't supported by Loki previously. 

Testing the new version of Loki out myself and will make a release if anyone else is still using Loki.

**Comment by jensbrak** [comment]
Updated my fork and made a release that works with The Bog Witch update.
Read the release notes before use. While it works for me and the files I've tested with, I cannot guarantee it works for everyone (ie I have not done much regression testing when it comes to older save files, even though it *should* work).
Let me know if you find any odd behavior or issues.

https://github.com/jensbrak/Loki/releases/tag/v0.16-beta

**Comment by jensbrak** [comment]
In case anyone, by chance, is still using Loki...
Updated my fork and made a release there that supports Call For Arms update (0.221.4).
As usual, read release notes and make back ups of your saves before trying it.
(I didn't test this one particularly well but load/save in Loki seems to work fine; Valheim does not complain or misbehave when using edited save.)

https://github.com/jensbrak/Loki/releases/tag/v0.17-beta

**Comment by Swirlyy** [comment]
It seems it is broken again, thanks to a new update..
Is there any way I can help keep Loki working and updated, I can code but you might have to show me the ropes to make it smoother?

**Comment by jensbrak** [comment]
> It seems it is broken again, thanks to a new update.. Is there any way I can help keep Loki working and updated, I can code but you might have to show me the ropes to make it smoother?

I can have a look and update my fork. It should be no problem. 
It's usually small things to fix to get it up and running again after an update.

(The thing that takes some time is to get some understanding on how Valheim works and to reverse engineer and analyze what the new version has changed. That is off topic here, but there are guides out there in case you want to dig into how it works.)

**Comment by jensbrak** [comment]
From what I gather, they only bumped the version number of the save file, but there's no actual changes in the format itself. Or so it seems, from a quick glance. The reason it fails is because Loki check the version and if it's a new(er) one it stops. Not sure why they bumped the version despite no changes, makes me a bit suspicious. 

There are a few new items and I've added those. Once I've verified that there's nothing too fishy about the updated version I'll upload it to my fork.

**Comment by jensbrak** [comment]
Out of sheer skepticism nothing but save version was changed I quickly did a line-by-line comparison of new and previous version of Valheim and how it saves/loads save files. Seems there is nothing but version changed actually. Quick testing of updated version of Loki seems to agree with that. 

Still, use caution and have a backup before using new version. Just in case. But here we go, here's Loki 0.18 with Five Years of Valheim update support, @Swirlyy and anyone else still using Loki.

https://github.com/jensbrak/Loki/releases/tag/v0.18-beta

**Comment by Mostwood** [comment]
@jensbrak Thank you for your hard work keeping this useful tool alive!
I was wondering if it would be possible to make cheater tag editable too? 
I'm running a server for a couple of friends and sometimes it crashes unexpectedly. Sometimes it happens during sea travel and all stuff gets lost so I have to spawn resources back for them. I heard they are going to introduce achievements in 1.0 and seems that my character of 7 years wouldn't be eligible with that tag on.

**Comment by jensbrak** [comment]
@Mostwood That'd prolly be a minor fix. I'm running Valheim servers myself and know the struggle. I've also moved into doing my own mods to Valheim, just for the fun of it. That being said, I will look into a fix that could fix that little thing. It's just a true/false value in the save file. Not sure they use it but I know... for whatever reason it's there it's annoying. I mean, if it's there in a save file to indicate you've messed with the save file - you're likely able to mess with the save file to erase it. I just omitted that fix out of sheer laziness. Keep you posted - and thanks! I do this for fun but fun increases tenfold when someone appreciates it, too.

**Comment by Mostwood** [comment]
@jensbrak Thank you so much! Have already starred your fork so I will be following any future updates anyway. After all it's not a big deal so don't sweat it if it's too much hustle. It is already a great tool as is.

**Comment by jensbrak** [comment]
@Mostwood - fixed it. The fix was easy but I was a bit rusty on that part of the code so I needed more than a couple of minutes to wrap my head around it. Still no big deal, fun little change :) 

I added a small button next to Used cheats: Yes/No that is enabled if cheater flag/tag is true. Once pressed, the tag is set to false AND stats counter for cheats is reset too. (the flag is set to true once any cheat command is used but then a counter is increased any succeeding time a cheat is used. I assume you'd want that counter to be reset as well?)

If that's not the wanted behavior let me know. I'll test it a bit more before making a release anyways.

**Comment by Mostwood** [comment]
@jensbrak Awesome! Huge thanks! Sorry, I somehow missed the notification and it took me a while to check my backlog of mail.

**Comment by jensbrak** [comment]
@Mostwood No worries, I waited to push the update until I heard from you. Now that I know you're still interested (take it as the implemented behavior is ok with you) I've made the release and bundled the build with it. You can find it here:

https://github.com/jensbrak/Loki/releases/tag/v0.18.1-beta

Let me know if there's any issues or oddities I might have missed!

**Comment by jensbrak** [comment]
Valheim 1.0 notice: As for the upcoming 1.0 release: I have no access to the 1.0 code required to make adjustments in Loki - and will not have until the release of Valheim 1.0 is actually made. And when it is, I will probably focus on playing the game rather than coding tools for it. We also don't know if devs will make an effort to actually encrypt the binaries to prevent reverse engineering or something like that. So any Loki for Valheim 1.0 may or may not happen and if it does, not sure when.

**Comment by jensbrak** [comment]
I did however, heh, write my own Character Editor for Valheim, completely from scratch. With a design that 99% focus on being easy to adopt for new Valheim versions. It still relies on Valheim 1.0 being possible to reverse engineer but now I have a code base 100% mine that is written with Valheim updates in mind. Currently doing final touches on a crude first version - but it can already do what Loki can and then some.

**Comment by jensbrak** [comment]
Ok. So 1.0 is out and it is... awesome. Right guys?! Hope you're busy playing it!

And. My own editor is almost done. It works with Valheim 1.0 already but there's a few minor wrinkles to iron out before I upload it - but it is solid in loading and saving 1.0-saves and goes a bit beyond what Loki did in terms of functionality.  So look out for `Norn` at my GitHub place, any day or even hour now! I don't wanna promote it here more than that, feels wrong. 

Thanks everyone cheering my attempts in keeping Loki alive a bit longer.  I did it for fun but now that has come to an end - but it's also what's encouraged me to write something of my own, from scratch. (Well that and my curiosity that never seem to stop)

_Above all: Thank you **Wufflez** for Loki and for inspiration and knowledge!_

**Comment by Wufflez** [comment]
Thankyou @jensbrak! I have been through a lot in life since playing Valheim and not had chance to play it for years let alone maintain a tool for it. Since releasing the tool i've got married, changed career and had my first child and @jensbrak has just kept this tool up to date like a hero. 
I do hope to get some free time soon, as the little one goes to nursery tomorrow for the first time and is *just* about starting to sleep better at night so maybe I'll get chance to play 1.0 soon and perhaps be inspired to update my own tool again, or perhaps help @jensbrak with his instead.
_Thanks again @jensbrak!_

**Comment by jensbrak** [comment]
FOUR times congratulations @Wufflez - for getting married, for having a kid, for changing career and for having your priorities in order focusing on that! I know the struggles, believe me, but I hope you get the chance to try out Valheim 1.0, it's pretty darn awesome!

I just pushed the first version of [Norn](https://github.com/jensbrak/Norn) now. Anyone wanna try it out feel free.  Edit: and now there's a [release ](https://github.com/jensbrak/Norn/releases) for it available, too. Let me know if anything is off, if you try it.
